### PR TITLE
Feat: kyber oracle

### LINKED ISSUE #10: DEX to accept others ERC20

### PR DESCRIPTION
**Pull request by nachomazzara**
Close #10 
- Kyber Token converter to swap ERC20 tokens to MANA

### Architecture

screen shot 2018-11-12 at 22 43 43

### KyberNetwork
- [getExpectedRate](https://developer.kyber.network/docs/KyberNetworkProxy/#getexpectedrate)
- [trade](https://developer.kyber.network/docs/KyberNetworkProxy/#trade)

### CODE REVIEW

File: contracts/auction/LANDAuction.sol [hunk 1]
```diff
@@ -18,31 +18,34 @@ contract LANDAuction is Ownable, LANDAuctionStorage {
...
 
         require(
-            _landRegistry.isContract(),
+            address(_landRegistry).isContract(),
             "The LANDRegistry token address must be a deployed contract"
         );
-        landRegistry = LANDRegistry(_landRegistry);
+        landRegistry = _landRegistry;
+
+        if (_dex != address(0)) {
```
Comment (donequis) [inline contracts/auction/LANDAuction.sol]: check's done inside setDex

File: contracts/auction/LANDAuction.sol [hunk 2]
```diff
@@ -190,11 +230,73 @@ contract LANDAuction is Ownable, LANDAuctionStorage {
...
+        uint256 _decimals, 
+        bool _shouldKeepToken) 
+    public onlyOwner 
+    {
+        require(
+            _address.isContract(),
+            "Tokens allowed should be a deployed ERC20 contract"
+        );
+        require(
+            _decimals > 0 && _decimals <= 18,
```
Comment (donequis) [inline contracts/auction/LANDAuction.sol]: <= MAX_DECIMALS ?

File: contracts/auction/LANDAuction.sol [hunk 3]
```diff
@@ -18,31 +18,34 @@ contract LANDAuction is Ownable, LANDAuctionStorage {
...
+        landRegistry = _landRegistry;
+
+        if (_dex != address(0)) {
+            setDex(_dex);
+        }
+
+        allowToken(address(_manaToken), 18, true);
+        manaToken = _manaToken;
 
         require(_initialPrice > _endPrice, "The start price should be greater than end price");
```
Comment (abarmat) [inline contracts/auction/LANDAuction.sol]: Maybe we could move this requires to the start of the function, under the first require()

Comment (nachomazzara) [reply contracts/auction/LANDAuction.sol]: I removed the `_initialPrice` and `_endPrice` in #16

File: contracts/auction/LANDAuction.sol [hunk 4]
```diff
@@ -210,4 +312,61 @@ contract LANDAuction is Ownable, LANDAuctionStorage {
...
     }
+
+    /**
+    * @dev Convert allowed token to MANA and transfer the change in MANA to the sender
+    * Note that we will use the slippageRate cause it has a 3% buffer
+    * @param _fromToken - ERC20 token to be converted
+    * @param _totalPrice - uint256 of the total amount in MANA
+    * @return bool to confirm the convertion was successfully
+    */
+    function convertSafe(ERC20 _fromToken, uint256 _totalPrice) internal returns (bool) {
```
Comment (abarmat) [inline contracts/auction/LANDAuction.sol]: Shouldn't we have a require() condition for the _fromToken being allowed?

File: contracts/auction/LANDAuction.sol [hunk 5]
```diff
@@ -210,4 +312,61 @@ contract LANDAuction is Ownable, LANDAuctionStorage {
...
+        uint256 prevBalance = manaToken.balanceOf(address(this));
+
+        uint256 tokenRate;
+        (, tokenRate) = dex.getExpectedRate(manaToken, _fromToken, _totalPrice);
+
+        uint256 totalPriceInToken = _totalPrice.mul(tokenRate).div(10 ** 18);
```
Comment (abarmat) [inline contracts/auction/LANDAuction.sol]: Maybe it's a good idea to have this as a function for the frontend to query the `totalPriceInToken` and show it to the user?

File: contracts/auction/LANDAuction.sol [hunk 6]
```diff
@@ -210,4 +312,61 @@ contract LANDAuction is Ownable, LANDAuctionStorage {
...
+
+        uint256 fromTokenDecimals = tokensAllowed[address(_fromToken)].decimals;
+        // Normalize to _fromToken decimals and calculate the amount of tokens to convert
+        if (MAX_DECIMALS > fromTokenDecimals)  {
+             // Ceil the result of the normalization always fue to convertions fee
```
Comment (abarmat) [inline contracts/auction/LANDAuction.sol]: `fue` (typo ?)

File: contracts/auction/LANDAuction.sol (hunk 5 above)
Comment (nachomazzara) [reply contracts/auction/LANDAuction.sol]: Have did something related [here](https://github.com/decentraland/land-auction/pull/15/files#diff-ac7aac0ba2190ba3d0ee5cbc786152fbR323)

**Pull request merged**